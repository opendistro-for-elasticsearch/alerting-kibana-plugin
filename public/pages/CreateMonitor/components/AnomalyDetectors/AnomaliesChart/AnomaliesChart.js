/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { EuiText, EuiSpacer } from '@elastic/eui';
import {
  Chart,
  getAxisId,
  Axis,
  getSpecId,
  LineSeries,
  niceTimeFormatter,
  Settings,
  Position,
  getAnnotationId,
  LineAnnotation,
} from '@elastic/charts';
import { ChartContainer } from '../../../../../components/ChartContainer/ChartContainer';
import DelayedLoader from '../../../../../components/DelayedLoader';

const getAxisTitle = (displayGrade, displayConfidence) => {
  if (displayGrade && displayConfidence) {
    return 'Anomaly grade / confidence';
  }
  return displayGrade ? 'Anomaly grade' : 'Anomaly confidence';
};

const AnomaliesChart = props => {
  const timeFormatter = niceTimeFormatter([props.startDateTime, props.endDateTime]);
  return (
    <DelayedLoader isLoading={props.isLoading}>
      {showLoader => (
        <React.Fragment>
          {props.showTitle ? (
            <EuiText size="xs">
              <strong>{props.title}</strong>
            </EuiText>
          ) : null}
          <EuiSpacer size="s" />
          <div>
            <ChartContainer
              style={{ height: '300px', width: '100%', opacity: showLoader ? 0.2 : 1 }}
            >
              <Chart>
                {props.showSettings ? (
                  <Settings
                    showLegend
                    legendPosition={Position.Bottom}
                    showLegendDisplayValue={false}
                  />
                ) : null}
                <Axis id={getAxisId('bottom')} position="bottom" tickFormat={timeFormatter} />
                <Axis
                  id={getAxisId('left')}
                  title={getAxisTitle(props.displayGrade, props.displayConfidence)}
                  position="left"
                  domain={{ min: 0, max: 1 }}
                />
                {props.annotationData ? (
                  <LineAnnotation
                    annotationId={getAnnotationId('anomalyAnnotation')}
                    domainType="yDomain"
                    dataValues={props.annotationData}
                    style={{
                      line: {
                        strokeWidth: 1.5,
                        stroke: '#f00',
                      },
                    }}
                  />
                ) : null}
                {props.displayGrade ? (
                  <LineSeries
                    id={getSpecId('Anomaly grade')}
                    xScaleType="time"
                    yScaleType="linear"
                    xAccessor={'plotTime'}
                    yAccessors={['anomalyGrade']}
                    data={props.anomalies}
                  />
                ) : null}
                {props.displayConfidence ? (
                  <LineSeries
                    id={getSpecId('Confidence')}
                    xScaleType="time"
                    yScaleType="linear"
                    xAccessor={'plotTime'}
                    yAccessors={['confidence']}
                    data={props.anomalies}
                  />
                ) : null}
              </Chart>
            </ChartContainer>
          </div>
        </React.Fragment>
      )}
    </DelayedLoader>
  );
};

AnomaliesChart.propTypes = {
  startDateTime: PropTypes.number,
  endDateTime: PropTypes.number,
  endDate: PropTypes.number,
  isLoading: PropTypes.bool,
  showTitle: PropTypes.bool,
  showSettings: PropTypes.bool,
  annotationData: PropTypes.array,
  anomalies: PropTypes.array.isRequired,
  title: PropTypes.string,
};

AnomaliesChart.defaultProps = {
  isLoading: false,
  showTitle: true,
  showSettings: true,
  anomalies: undefined,
  title: 'Sample preview for anomaly score',
};

export { AnomaliesChart };
