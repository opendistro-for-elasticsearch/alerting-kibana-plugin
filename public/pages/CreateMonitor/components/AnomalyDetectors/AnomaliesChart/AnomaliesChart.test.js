/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { AnomaliesChart } from './AnomaliesChart';
import moment from 'moment';

const startTime = moment('2018-10-25T09:30:00').valueOf();
const endTime = moment('2018-10-29T09:30:00').valueOf();

describe('AnomaliesChart', () => {
  test('renders ', () => {
    const sampleData = [
      {
        anomalyGrade: 0.9983687181753063,
        anomalyScore: 0.8381447468893426,
        confidence: 0.42865659282252266,
        detectorId: 'temp',
        endTime: 1569097677667,
        plotTime: 1569097377667,
        startTime: 1569097077667,
      },
    ];
    const component = (
      <AnomaliesChart
        anomalies={sampleData}
        startDateTime={startTime}
        endDateTime={endTime}
        isLoading={false}
        displayGrade
        displayConfidence
        showTitle
      />
    );
    expect(render(component)).toMatchSnapshot();
  });
});
