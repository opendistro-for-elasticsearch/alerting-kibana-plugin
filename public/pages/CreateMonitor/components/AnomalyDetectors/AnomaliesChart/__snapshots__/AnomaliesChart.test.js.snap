// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnomaliesChart renders  1`] = `
Array [
  <div
    class="euiText euiText--extraSmall"
  >
    <strong>
      Sample preview for anomaly score
    </strong>
  </div>,
  <div
    class="euiSpacer euiSpacer--s"
  />,
  <div>
    <div
      style="border-radius:5px;padding:10px;border:1px solid #D9D9D9;height:300px;width:100%;opacity:1"
    >
      <div
        class="echChart"
      >
        <div
          class="echContainer"
        >
          <div
            class="echChartResizer"
          />
          <div
            class="echCrosshair"
          />
          <div
            class="echTooltip echTooltip--hidden"
          />
          <div
            class="echAnnotation__tooltip echAnnotation__tooltip--hidden"
          />
          <svg
            class="echHighlighter"
          >
            <g
              transform="translate(0, 0) rotate(0)"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
]
`;
